File: src/main/java/io/disassemble/javanalysis/util/InsnUtil.java
package io.disassemble.javanalysis.util;

import io.disassemble.javanalysis.insn.*;

import java.util.Arrays;

import static javassist.bytecode.Opcode.*;

/**
 * @author Tyler Sedlar
 * @since 5/20/2017
 */
public class InsnUtil {

    public static boolean isLdcInsn(int opcode) {
        return opcode == LDC || opcode == LDC_W || opcode == LDC2_W;
    }

    public static boolean isIntInsn(int opcode) {
        return opcode == BIPUSH || opcode == SIPUSH || opcode == NEWARRAY;
    }

    public static boolean isVarInsn(int opcode) {
        return (opcode >= ILOAD && opcode <= ALOAD_3) || (opcode >= ISTORE && opcode <= ASTORE_3) || opcode == RET;
    }

    public static boolean isJumpInsn(int opcode) {
        return (opcode >= IFEQ && opcode <= JSR) || (opcode >= IFNULL && opcode <= JSR_W);
    }

    public static boolean isIncrementInsn(int opcode) {
        return opcode == IINC;
    }

    public static boolean isFieldInsn(int opcode) {
        return opcode >= GETSTATIC && opcode <= PUTFIELD;
    }

    public static boolean isMethodInsn(int opcode) {
        return opcode >= INVOKEVIRTUAL && opcode <= INVOKEINTERFACE;
    }

    public static boolean isInvokeDynamicInsn(int opcode) {
        return opcode == INVOKEDYNAMIC;
    }

    public static boolean isTypeInsn(int opcode) {
        return opcode == ANEWARRAY || opcode == CHECKCAST || opcode == MULTIANEWARRAY;
    }

    public static boolean isTableSwitchInsn(int opcode) {
        return opcode == TABLESWITCH;
    }

    public static boolean isLookupSwitchInsn(int opcode) {
        return opcode == LOOKUPSWITCH;
    }

    public static boolean isUnderscoreInsn(int opcode) {
        return (opcode >= ILOAD_0 && opcode <= ALOAD_3) || (opcode >= ISTORE_0 && opcode <= ASTORE_3) ||
                (opcode >= ICONST_M1 && opcode <= DCONST_1);
    }

    public static int underVal(int opcode) {
        switch(opcode) {
            case ILOAD_0:
            case LLOAD_0:
            case FLOAD_0:
            case DLOAD_0:
            case ALOAD_0:
            case ISTORE_0:
            case LSTORE_0:
            case FSTORE_0:
            case DSTORE_0:
            case ASTORE_0:
            case ICONST_0:
            case LCONST_0:
            case FCONST_0:
            case DCONST_0: {
                return 0;
            }
            case ILOAD_1:
            case LLOAD_1:
            case FLOAD_1:
            case DLOAD_1:
            case ALOAD_1:
            case ISTORE_1:
            case LSTORE_1:
            case FSTORE_1:
            case DSTORE_1:
            case ASTORE_1:
            case ICONST_1:
            case LCONST_1:
            case FCONST_1:
            case DCONST_1: {
                return 1;
            }
            case ILOAD_2:
            case LLOAD_2:
            case FLOAD_2:
            case DLOAD_2:
            case ALOAD_2:
            case ISTORE_2:
            case LSTORE_2:
            case FSTORE_2:
            case DSTORE_2:
            case ASTORE_2:
            case ICONST_2:
            case FCONST_2: {
                return 2;
            }
            case ILOAD_3:
            case LLOAD_3:
            case FLOAD_3:
            case DLOAD_3:
            case ALOAD_3:
            case ISTORE_3:
            case LSTORE_3:
            case FSTORE_3:
            case DSTORE_3:
            case ASTORE_3:
            case ICONST_3: {
                return 3;
            }
            case ICONST_4: {
                return 4;
            }
            case ICONST_5: {
                return 5;
            }
            default: {
                return -1;
            }
        }
    }

    public static String stringify(CtInsn insn) {
        String label = "";
        Object data = "";
        if (insn instanceof LdcInsn) {
            label = "cst";
            data = ((LdcInsn) insn).cst();
        } else if (insn instanceof IntInsn) {
            label = "operand";
            data = ((IntInsn) insn).operand();
        } else if (insn instanceof VarInsn) {
            label = "var";
            data = ((VarInsn) insn).var();
        } else if (insn instanceof JumpInsn) {
            label = "target";
            data = ((JumpInsn) insn).target();
        } else if (insn instanceof ClassMemberInsn) {
            label = (insn instanceof FieldInsn ? "field" : "method");
            data = ((ClassMemberInsn) insn).key();
        } else if (insn instanceof TypeInsn) {
            label = "type";
            data = ((TypeInsn) insn).type();
        } else if (insn instanceof TableSwitchInsn) {
            label = "data";
            TableSwitchInsn table = (TableSwitchInsn) insn;
            data = ("low = " + table.low() + ", high = " + table.high() + ", defaultIndex = " +
                    table.defaultIndex() + ", indices = " + Arrays.toString(table.indices()));
        } else if (insn instanceof LookupSwitchInsn) {
            label = "data";
            LookupSwitchInsn table = (LookupSwitchInsn) insn;
            data = ("defaultIndex = " + table.defaultIndex() + ", keys = " + Arrays.toString(table.keys()) +
                    ", indices = " + Arrays.toString(table.indices()));
        }
        String output = insn.index() + ": " + insn.opname();
        if (!label.isEmpty()) {
            output += (" (" + label + ": " + data + ")");
        }
        return output;
    }
}

File: src/main/java/io/disassemble/javanalysis/util/InsnUtil.kt
package io.disassemble.javanalysis.util

import io.disassemble.javanalysis.insn.*
import javassist.bytecode.Opcode.*

/**
 * @author Tyler Sedlar
 * @since 5/20/2017
 */
object InsnUtil {

    /**
     * Checks if the given opcode is an [LdcInsn].
     *
     * @param opcode The opcode to check against.
     *
     * @return true if the given opcode is an [LdcInsn], otherwise false.
     */
    fun isLdcInsn(opcode: Int): Boolean {
        return opcode == LDC || opcode == LDC_W || opcode == LDC2_W
    }

    /**
     * Checks if the given opcode is an [IntInsn].
     *
     * @param opcode The opcode to check against.
     *
     * @return true if the given opcode is an [IntInsn], otherwise false.
     */
    fun isIntInsn(opcode: Int): Boolean {
        return opcode == BIPUSH || opcode == SIPUSH || opcode == NEWARRAY
    }

    /**
     * Checks if the given opcode is a [VarInsn].
     *
     * @param opcode The opcode to check against.
     *
     * @return true if the given opcode is a [VarInsn], otherwise false.
     */
    fun isVarInsn(opcode: Int): Boolean {
        return opcode in ILOAD..ALOAD_3 || opcode in ISTORE..ASTORE_3 || opcode == RET
    }

    /**
     * Checks if the given opcode is a [JumpInsn].
     *
     * @param opcode The opcode to check against.
     *
     * @return true if the given opcode is a [JumpInsn], otherwise false.
     */
    fun isJumpInsn(opcode: Int): Boolean {
        return opcode in IFEQ..JSR || opcode in IFNULL..JSR_W
    }

    /**
     * Checks if the given opcode is a [IncrementInsn].
     *
     * @param opcode The opcode to check against.
     *
     * @return true if the given opcode is a [IncrementInsn], otherwise false.
     */
    fun isIncrementInsn(opcode: Int): Boolean {
        return opcode == IINC
    }

    /**
     * Checks if the given opcode is a [FieldInsn].
     *
     * @param opcode The opcode to check against.
     *
     * @return true if the given opcode is a [FieldInsn], otherwise false.
     */
    fun isFieldInsn(opcode: Int): Boolean {
        return opcode in GETSTATIC..PUTFIELD
    }

    /**
     * Checks if the given opcode is a [MethodInsn].
     *
     * @param opcode The opcode to check against.
     *
     * @return true if the given opcode is a [MethodInsn], otherwise false.
     */
    fun isMethodInsn(opcode: Int): Boolean {
        return opcode in INVOKEVIRTUAL..INVOKEINTERFACE
    }

    /**
     * Checks if the given opcode is an [InvokeDynamicInsn].
     *
     * @param opcode The opcode to check against.
     *
     * @return true if the given opcode is an [InvokeDynamicInsn], otherwise false.
     */
    fun isInvokeDynamicInsn(opcode: Int): Boolean {
        return opcode == INVOKEDYNAMIC
    }

    /**
     * Checks if the given opcode is a [TypeInsn].
     *
     * @param opcode The opcode to check against.
     *
     * @return true if the given opcode is a [TypeInsn], otherwise false.
     */
    fun isTypeInsn(opcode: Int): Boolean {
        return opcode == ANEWARRAY || opcode == CHECKCAST || opcode == MULTIANEWARRAY
    }

    /**
     * Checks if the given opcode is a [TableSwitchInsn].
     *
     * @param opcode The opcode to check against.
     *
     * @return true if the given opcode is a [TableSwitchInsn], otherwise false.
     */
    fun isTableSwitchInsn(opcode: Int): Boolean {
        return opcode == TABLESWITCH
    }

    /**
     * Checks if the given opcode is a [LookupSwitchInsn].
     *
     * @param opcode The opcode to check against.
     *
     * @return true if the given opcode is a [LookupSwitchInsn], otherwise false.
     */
    fun isLookupSwitchInsn(opcode: Int): Boolean {
        return opcode == LOOKUPSWITCH
    }

    /**
     * Checks if the given opcode is an opcode containing an underscore (iload_0, iload_1, etc.)
     *
     * @param opcode The opcode to check against.
     *
     * @return true if the given opcode is an ontaining an underscore (iload_0, iload_1, etc.)
     */
    fun isUnderscoreInsn(opcode: Int): Boolean {
        return opcode in ILOAD_0..ALOAD_3 || opcode in ISTORE_0..ASTORE_3 ||
                opcode in ICONST_M1..DCONST_1
    }

    /**
     * Gets the value of the underscore variable. (iload_0 = 0, iload_1 = 1, etc.)
     *
     * @param opcode The opcode to get the underscore value from.
     *
     * @return The value of the underscore variable. (iload_0 = 0, iload_1 = 1, etc.)
     */
    fun underVal(opcode: Int): Int {
        when (opcode) {
            ILOAD_0, LLOAD_0, FLOAD_0, DLOAD_0, ALOAD_0, ISTORE_0, LSTORE_0,
            FSTORE_0, DSTORE_0, ASTORE_0, ICONST_0, LCONST_0, FCONST_0, DCONST_0 -> {
                return 0
            }
            ILOAD_1, LLOAD_1, FLOAD_1, DLOAD_1, ALOAD_1, ISTORE_1, LSTORE_1,
            FSTORE_1, DSTORE_1, ASTORE_1, ICONST_1, LCONST_1, FCONST_1, DCONST_1 -> {
                return 1
            }
            ILOAD_2, LLOAD_2, FLOAD_2, DLOAD_2, ALOAD_2, ISTORE_2, LSTORE_2,
            FSTORE_2, DSTORE_2, ASTORE_2, ICONST_2, FCONST_2 -> {
                return 2
            }
            ILOAD_3, LLOAD_3, FLOAD_3, DLOAD_3, ALOAD_3, ISTORE_3, LSTORE_3,
            FSTORE_3, DSTORE_3, ASTORE_3, ICONST_3 -> {
                return 3
            }
            ICONST_4 -> {
                return 4
            }
            ICONST_5 -> {
                return 5
            }
            else -> {
                return -1
            }
        }
    }

    /**
     * Gets the string representation of the given [CtInsn].
     *
     * @param insn The [CtInsn] to get a visual representation for.
     *
     * @return The string representation of the given [CtInsn].
     */
    fun stringify(insn: CtInsn): String {
        var label = ""
        var data: Any = ""
        when (insn) {
            is LdcInsn -> {
                label = "cst"
                data = insn.cst
            }
            is IntInsn -> {
                label = "operand"
                data = insn.operand
            }
            is VarInsn -> {
                label = "var"
                data = insn.variable
            }
            is JumpInsn -> {
                label = "target"
                data = insn.target
            }
            is ClassMemberInsn -> {
                label = if (insn is FieldInsn) "field" else "method"
                data = insn.key
            }
            is TypeInsn -> {
                label = "type"
                data = insn.type
            }
            is TableSwitchInsn -> {
                label = "data"
                data = "low = " + insn.low + ", high = " + insn.high + ", defaultIndex = " +
                        insn.defaultIndex + ", indices = " + insn.indices.contentToString()
            }
            is LookupSwitchInsn -> {
                label = "data"
                data = "defaultIndex = " + insn.defaultIndex + ", keys = " + insn.keys.contentToString() +
                        ", indices = " + insn.indices.contentToString()
            }
            is IncrementInsn -> {
                label = "data"
                data = "var = " + insn.variable + ", incr = " + insn.increment
            }
        }
        var output = insn.opname
        if (label.isNotEmpty()) {
            output += " ($label: $data)"
        }
        return output
    }
}
